capsules/extra/src/lps22hb.rs

```diff
@@ -0,0 +1,210 @@
+//! > The device comprises a sensing element and an IC interface 
+//! > which communicates through I2C or SPI from the sensing element 
+//! > to the application.
+//!
+//! Driver Semantics
+//! ----------------
+//!
+//! This driver exposes the LPS22HB's pressure functionality via
+//! the [PressureDriver] HIL interface. It doesn't support handling
+//! multiple concurrent pressure requests.
+//!
+//! Usage
+//! -----
+//!
+//! ```rust
+//! # use kernel::static_init;
+//!
+//! let lps22hb_i2c = static_init!(
+//!     capsules::virtual_i2c::I2CDevice,
+//!     capsules::virtual_i2c::I2CDevice::new(i2c_bus, 0x5C));
+//! let lps22hb = static_init!(
+//!     capsules::lps22hb::Lps22hb<'static>,
+//!     capsules::lps22hb::Lps22hb::new(lps22hb_i2c,
+//!         &mut capsules::lps22hb::BUFFER));
+//! lps22hb_i2c.set_client(lps22hb);
+//! ```
+
+use core::cell::Cell;
+use kernel::hil::i2c::{self, I2CClient, I2CDevice};
+use kernel::hil::sensors::{PressureClient, PressureDriver};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::ErrorCode;
+
+/// Register values
+
+const REGISTER_AUTO_INCREMENT: u8 = 0x80;
+const CTRL_REG1_POWER_ON: u8 = 0x50;
+const CTRL_REG1_POWER_OFF: u8 = 0x00;
+
+#[allow(dead_code)]
+enum Registers {
+    IntCfgReg = 0x0B,
+    ThsPL = 0x0C,
+    ThsPH = 0x0D,
+    WhoAmI = 0x0F,
+    CtrlReg1 = 0x10,
+    CtrlReg2 = 0x11,
+    CtrlReg3 = 0x12,
+    FifoCtrl = 0x14,
+    RefPXl = 0x15,
+    RefPL = 0x16,
+    RefPH = 0x17,
+    RpdsL = 0x18,
+    RpdsH = 0x19,
+    ResConf = 0x1A,
+    IntSourceReg = 0x25,
+    FifoStatus = 0x26,
+    StatusReg = 0x27,
+    PressOutXl = 0x28,
+    PressOutL = 0x29,
+    PressOutH = 0x2A,
+    TempOutL = 0x2B,
+    TempOutH = 0x2C,
+    LpfpRes = 0x33
+}
+
+pub struct Lps22hb<'a, I: I2CDevice> {
+    buffer: TakeCell<'static, [u8]>,
+    i2c_bus: &'a I,
+    pressure_client: OptionalCell<&'a dyn PressureClient>,
+    pending_pressure: Cell<bool>,
+    state: Cell<State>
+}
+
+impl<'a, I: I2CDevice> Lps22hb<'a, I> {
+    pub fn new(
+        i2c_bus: &'a I,
+        buffer: &'static mut [u8],
+    ) -> Lps22hb<'a, I> {
+        Lps22hb { 
+            buffer: TakeCell::new(buffer), 
+            i2c_bus: i2c_bus,
+            pressure_client: OptionalCell::empty(), 
+            pending_pressure: Cell::new(false), 
+            state: Cell::new(State::Idle)
+        }
+    }
+
+    fn start_measurement(&self) -> Result<(), ErrorCode> {
+        self.buffer
+            .take()
+            .map(|buffer| {
+                self.i2c_bus.enable();
+                match self.state.get() {
+                    State::Idle => {
+                        buffer[0] = Registers::CtrlReg1 as u8;
+                        buffer[1] = CTRL_REG1_POWER_ON;
+
+                        if let Err((_error, buffer)) = self.i2c_bus.write(buffer, 2) {
+                            self.buffer.replace(buffer);
+                            self.i2c_bus.disable();
```

Comment:
In this case the `start_measurement()` command should return something other than BUSY.